kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```
@@ -0,0 +1,552 @@
+			if isSubdirectory(filepath.Join(projectRoot, excludeDir), filepath.Dir(path)) {
+				return false, filepath.SkipDir
+			}
+		}
+		return false, nil
+	}
+
+	// only index .rs files for now. It's possible that some rust files depend on non rust files, like if they use include_bytes,
+	// TODO(wittrock): figure out a better scheme for including all actually depended-on files and still exclude files like .deps and .rmeta
+	if filepath.Ext(path) != ".rs" {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+type collectCrateSourcesFunc func(ctx context.Context, e *extractor, sourceDirs source) ([]string, []*apb.CompilationUnit_FileInput, error)
+
+// getSourceFiles walks the file system based on the provided sourceDirs,
+// identifies relevant source files (currently only .rs files), adds them to the
+// kzip archive, and returns lists of their relative paths and FileInput protos.
+func collectCrateSourcesImpl(ctx context.Context, e *extractor, sourceDirs source) ([]string, []*apb.CompilationUnit_FileInput, error) {
+	var sourceFiles []string
+	var requiredInputs []*apb.CompilationUnit_FileInput
+
+	// cache required inputs for each include_dir
+	for _, includeDir := range sourceDirs.IncludeDirs {
+		if e.requiredInputsCache.inputsCache[includeDir] != nil {
+			requiredInputs = append(requiredInputs, e.requiredInputsCache.inputsCache[includeDir]...)
+			sourceFiles = append(sourceFiles, e.requiredInputsCache.sourceFilesCache[includeDir]...)
+
+			// we don't need to add this file to the kzip_writer, because we know it's already been written.
+			continue
+		}
+
+		dirRequiredInputs := []*apb.CompilationUnit_FileInput{}
+		err := vfs.Walk(ctx, filepath.Join(e.projectRoot, includeDir), func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			shouldInclude, err := shouldIncludeFile(info, path, sourceDirs, e.projectRoot)
+			if err != nil || !shouldInclude {
+				return nil
```

log rather than silently returning error?
I suppose if there are lots of data files in these directories, maybe this would be spam.
Up to you.

---

We do practically have a bunch of these files in Fuchsia unfortunately, so I'd like to not log here.